diesel/src/sqlite/expression/functions.rs

```diff
@@ -959,4 +959,136 @@ extern "SQL" {
     #[sql_name = "json_quote"]
     #[cfg(feature = "sqlite")]
     fn json_quote<J: SqlType + SingleValue>(j: J) -> Json;
+
+    /// The json_patch(T,P) SQL function runs the RFC-7396 MergePatch algorithm to apply patch P against input T. The patched copy of T is returned.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # include!("../../doctest_setup.rs");
+    /// #
+    /// # fn main() {
+    /// #     #[cfg(feature = "serde_json")]
+    /// #     run_test().unwrap();
+    /// # }
+    /// #
+    /// # #[cfg(feature = "serde_json")]
+    /// # fn run_test() -> QueryResult<()> {
+    /// #     use diesel::dsl::{sql, json_patch};
+    /// #     use serde_json::{json, Value};
+    /// #     use diesel::sql_types::{Text, Json, Nullable};
+    /// #     let connection = &mut establish_connection();
+    ///
+    /// let version = diesel::select(sql::<Text>("sqlite_version();"))
+    ///         .get_result::<String>(connection)?;
+    ///
+    /// // Querying SQLite version should not fail.
+    /// let version_components: Vec<&str> = version.split('.').collect();
+    /// let major: u32 = version_components[0].parse().unwrap();
+    /// let minor: u32 = version_components[1].parse().unwrap();
+    /// let patch: u32 = version_components[2].parse().unwrap();
+    ///
+    /// if major > 3 || (major == 3 && minor >= 38) {
+    ///     /* Valid sqlite version, do nothing */
+    /// } else {
+    ///     println!("SQLite version is too old, skipping the test.");
+    ///     return Ok(());
+    /// }
+    /// let result = diesel::select(json_patch::<Json, _, _>(json!({"a":1,"b":2}), json!({"c":3,"d":4})))
+    ///     .get_result::<Value>(connection)?;
+    ///
+    /// assert_eq!(json!({"a":1, "b":2, "c":3, "d":4}), result);
+    ///
+    /// let result = diesel::select(json_patch::<Json, _, _>(json!({"a":[1,2],"b":2}), json!({"a":9})))
+    ///     .get_result::<Value>(connection)?;
+    ///
+    /// assert_eq!(json!({"a":9,"b":2}), result);
+    ///
+    /// let result = diesel::select(json_patch::<Json, _, _>(json!({"a":[1,2],"b":2}), json!({"a":null})))
+    ///     .get_result::<Value>(connection)?;
+    ///
+    /// assert_eq!(json!({"b":2}), result);
+    ///
+    /// let result = diesel::select(json_patch::<Json, _, _>(json!({"a":1,"b":2}), json!({"a":9,"b":null,"c":8})))
+    ///     .get_result::<Value>(connection)?;
+    ///
+    /// assert_eq!(json!({"a":9,"c":8}), result);
+    ///
+    /// let result = diesel::select(json_patch::<Json, _, _>(json!({"a":{"x":1,"y":2},"b":3}), json!({"a":{"y":9},"c":8})))
+    ///     .get_result::<Value>(connection)?;
+    ///
+    /// assert_eq!(json!({"a":{"x":1,"y":9},"b":3,"c":8}), result);
+    ///
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[sql_name = "json_patch"]
+    #[cfg(feature = "sqlite")]
+    fn json_patch<J: JsonOrNullableJson + SingleValue>(j: J, patch: J) -> Json;
```

Comment:
The return type is not correct. This function (and also `jsonb_patch`) returns `Null` if either the first or second argument is nullable.

We might want something like 
```suggestion
    fn json_patch<J1: JsonOrNullableJson + SingleValue + CombinedNullableValue<J2, Json>, J2: JsonOrNullableJson + SingleValue>(j: J1, patch: J1) -> J1::Out;
```

with the `CombinedNullableValue` trait defined here:

https://github.com/diesel-rs/diesel/blob/c5cdaa4622309784cd57ec446c62a3e5d309cae0/diesel/src/pg/expression/expression_methods.rs#L3936-L3949 

(That needs to be moved to `diesel/src/expressions/nullable.rs` into a `pub(crate) private` module.